dali/core/mm/default_resources.cc

```diff
@@ -0,0 +1,282 @@
+#include "dali/core/format.h"
+#include "dali/core/mm/malloc_resource.h"
+#include "dali/core/device_guard.h"
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/mm/composite_resource.h"
+#include "rmm/mr/device/owning_wrapper.hpp"
+#include "rmm/mr/device/cuda_memory_resource.hpp"
+#include "rmm/mr/device/pool_memory_resource.hpp"
+
+namespace dali {
+namespace mm {
+
+namespace {
+
+template <typename T>
+inline std::shared_ptr<T> wrap(T *p, bool own) {
+  if (own)
+    return std::shared_ptr<T>(p);
+  else
+    return std::shared_ptr<T>(p, [](T*){});
+}
+
+struct DefaultResources {
+  ~DefaultResources() {
+    ReleasePinned();
+    ReleaseDevice();
+    ReleaseHost();
+  }
+
+  static DefaultResources &instance() {
+    static DefaultResources dr;
+    return dr;
+  }
+
+  std::shared_ptr<host_memory_resource> host;
+  std::shared_ptr<pinned_async_resource> pinned_async;
+  std::vector<std::shared_ptr<device_async_resource>> device;
+
+  void ReleasePinned() {
+    Release(pinned_async);
+  }
+
+  void ReleaseDevice() {
+    Release(device);
+  }
+
+  void ReleaseHost() {
+    Release(host);
+  }
+
+ private:
+  template <typename Ptr>
+  void Release(Ptr &p) noexcept(false) {
+    if (cudaGetLastError() == cudaErrorCudartUnloading) {
+      abandon(p);
+    } else {
+      p = {};
+    }
+  }
+
+  template <typename T>
+  void abandon(std::vector<T> &v) {
+    for (auto &x : v)
+      abandon(x);
+    v.clear();
+  }
+
+  template <typename T>
+  void abandon(std::shared_ptr<T> &p) {
+    // shared_ptr doesn't have a release() function like one in unique_ptr so we could
+    // safely abandon the pointer.
+    // What happens here: we create (by placement new) a shared pointer and move the original
+    // pointer there - this way the original pointer is released and the new pointer
+    // is allocated dynamically on stack and never destroyed.
+    // This accomplishes the intentional leaking of the pointer.
+    using Ptr = std::shared_ptr<T>;
+    std::aligned_storage_t<sizeof(Ptr), alignof(Ptr)> dump;
+    new (&dump) Ptr(std::move(p));
+  }
+};
+
+#define g_resources (DefaultResources::instance())
+
+
+inline std::shared_ptr<host_memory_resource> CreateDefaultHostResource() {
+  return std::make_shared<malloc_memory_resource>();
+}
+
+struct CUDARTLoader {
+  CUDARTLoader() {
+    int device_id = 0;
+    CUDA_CALL(cudaGetDevice(&device_id));
+    DeviceGuard dg(device_id);
+  }
+};
+
+
+template <class Callable>
+struct CallAtExit {
+  explicit CallAtExit(Callable &&c) : callable(std::move(c)) {}
+  ~CallAtExit() {
+    callable();
+  }
+  Callable callable;
+};
+
+template <typename Callable>
+CallAtExit<Callable> AtExit(Callable &&c) {
+  return CallAtExit<Callable>(std::forward<Callable>(c));
+}
+
+inline std::shared_ptr<device_async_resource> CreateDefaultDeviceResource() {
+  static CUDARTLoader CUDAInit;
+#ifdef DALI_USE_RMM_DEVICE_RESOURCE
+  auto upstream = std::make_shared<rmm::mr::cuda_memory_resource>();
```

Comment:
Is the plan to keep only the RMM implementation in the future?

Reply:
Another question. In the non-rmm implementation you made the upstream resource static. Why not here?

Reply:
1. Whether we're going to use RMM _implementation_ (as opposed to interface) is a matter of benchmarking.
2. RMM is more sensitive than our code to errors in destructors, so I play it safe here.